test/controller.test.ts

```diff
@@ -1981,6 +1966,536 @@ contract('Controller', ([owner, accountOwner1, accountOperator1, random]) => {
+      )
+      assert.equal(
+        senderBalanceBefore.minus(senderBalanceAfter).toString(),
+        collateralToDeposit.toString(),
+        'Sender collateral asset balance mismatch',
+      )
+      assert.equal(vaultAfter.collateralAssets.length, 1, 'Vault collateral asset array length mismatch')
+      assert.equal(vaultAfter.shortOtokens.length, 1, 'Vault short otoken array length mismatch')
+      assert.equal(
+        vaultAfter.collateralAssets[0],
+        usdc.address,
+        'Collateral asset address deposited into vault mismatch',
+      )
+      assert.equal(
+        vaultAfter.shortOtokens[0],
+        shortOtoken.address,
+        'Short otoken address deposited into vault mismatch',
+      )
+      assert.equal(
+        senderShortBalanceAfter.minus(senderShortBalanceBefore).toString(),
+        amountToMint.toString(),
+        'Short otoken amount minted mismatch',
+      )
+
+      // transger minted short ototken to hodler`
+      shortOtoken.transfer(holder1, amountToMint, {from: accountOwner1})
+    })
+
+    it('should revert exercising un-expired otoken', async () => {
+      const shortAmountToBurn = new BigNumber('1')
+      const actionArgs = [
+        {
+          actionType: ActionType.Exercise,
+          owner: ZERO_ADDR,
+          sender: holder1,
+          asset: shortOtoken.address,
+          vaultId: '0',
+          amount: shortAmountToBurn.toNumber(),
+          index: '0',
+          data: ZERO_ADDR,
+        },
+      ]
+
+      assert.equal(await controller.isExpired(shortOtoken.address), false, 'Short otoken is already expired')
+
+      await expectRevert(
+        controller.operate(actionArgs, {from: holder1}),
+        'Controller: can not exercise un-expired otoken',
+      )
+    })
+
+    it('should revert exercising after expiry, when price is not finalized yet', async () => {
+      // past time after expiry
+      await time.increase(60 * 61 * 24) // increase time with one hour in seconds
+      // set price in Oracle Mock, 150$ at expiry, expire ITM
+      await oracle.setExpiryPrice(
+        await shortOtoken.underlyingAsset(),
+        new BigNumber(await shortOtoken.expiryTimestamp()),
+        new BigNumber(150).times(new BigNumber(10).exponentiatedBy(18)),
+      )
+      // set it as not finalized in mock
+      await oracle.setIsFinalized(
+        await shortOtoken.underlyingAsset(),
+        new BigNumber(await shortOtoken.expiryTimestamp()),
+        false,
+      )
+
+      const shortAmountToBurn = new BigNumber('1')
+      const actionArgs = [
+        {
+          actionType: ActionType.Exercise,
+          owner: ZERO_ADDR,
+          sender: holder1,
+          asset: shortOtoken.address,
+          vaultId: '0',
+          amount: shortAmountToBurn.toNumber(),
+          index: '0',
+          data: ZERO_ADDR,
+        },
+      ]
+
+      assert.equal(await controller.isExpired(shortOtoken.address), true, 'Short otoken is not expired yet')
```

Comment:
This message should be price not finalized

Reply:
mmm this assert check that the Otoken is expired! can u explain more ?